### PR TITLE
Revert "Fix #3485, #3540: Don't use dropout for predicting test sets"

### LINKED ISSUE #3485: Inconsistent results with dart booster
I am getting inconsistent predictions using dart booster if I save it to file and then load again.
Using other boosters I have consistent predictions and all work as should.

I am using:
- Linux 4.17.5-1-ARCH
- Python 3.6.6
- xgboost 0.72.1
- numpy 1.14.5

```
import xgboost as xgb
import numpy as np

data = np.random.rand(10,1)
labels = np.random.rand(10,1) > 0.5
train = xgb.DMatrix(data, label=labels)
predict = xgb.DMatrix(np.random.rand(1,1))

param = {'booster': 'dart', 
         'silent':'1'}
bst = xgb.Booster()
bst = xgb.train(param.items(), train)
bst.save_model("xgb_test.bin")

for i in range(0,10): #run load model and prediction ten times to see difference
    bst = xgb.Booster()
    bst.load_model('xgb_test.bin')
    new_value = bst.predict(predict)
    print(new_value)
    if i == 0:
        last_value = new_value
    if last_value != new_value:
        raise SystemExit
```

### PR DESCRIPTION
**Pull request by hcho3**
Reverts dmlc/xgboost#3556, as it breaks the logic for custom evaluation and custom objective. I've documented the use of `ntree_limit` when predicting with test sets.

### CODE REVIEW

**Comment by khotilov** [comment]
Perhaps, here https://xgboost.readthedocs.io/en/latest/parameter.html#additional-parameters-for-dart-booster-booster-dart it would be a good place to add a sentence briefly explaining how predict works for DART. While it's not a "booster parameter" per se, it's an important parameter to use this booster properly.

**Comment by hcho3** [comment]
@khotilov Not sure if a sentence would be sufficient. How about adding
```rst
.. note:: Using ``predict()`` with DART booster

  If the booster object is DART type, ``predict()`` will perform dropouts, i.e. only
  some of the trees will be evaluated. This will produce incorrect results if ``data`` is
  not the training data. To obtain correct results on test sets, set ``ntree_limit`` to
  a nonzero value, e.g.

    .. code-block:: python

      preds = bst.predict(dtest, ntree_limit=num_round)
```

**Comment by RAMitchell** [comment]
Another way of dealing with this problem would be to cache the pointer to the training matrix and only use dropout when PredictBatch is called with the training matrix. This would automatically produce "correct" results without need for users to modify parameters. This behaviour would have to be carefully documented.

**Comment by hcho3** [comment]
@RAMitchell For 0.80 release, I think we should retain the existing behavior. Let's come back to it after the release, to come up with a better design for DART booster.
